R/tm_a_pca.R

```diff
@@ -1132,30 +1183,43 @@ srv_a_pca <- function(id, data, reporter, filter_panel_api, dat, plot_height, pl
       title = "R Code for PCA"
     )
 
-    ### REPORTER
-    if (with_reporter) {
-      card_fun <- function(comment, label) {
-        card <- teal::report_card_template(
-          title = "Principal Component Analysis Plot",
-          label = label,
-          with_filter = with_filter,
-          filter_panel_api = filter_panel_api
-        )
-        card$append_text("Principal Components Table", "header3")
-        card$append_table(computation()[["tbl_importance"]])
-        card$append_text("Eigenvectors Table", "header3")
-        card$append_table(computation()[["tbl_eigenvector"]])
-        card$append_text("Plot", "header3")
-        card$append_plot(plot_r(), dim = pws$dim())
-        if (!comment == "") {
-          card$append_text("Comment", "header3")
-          card$append_text(comment)
-        }
-        card$append_src(source_code_r())
-        card
-      }
-      teal.reporter::simple_reporter_srv("simple_reporter", reporter = reporter, card_fun = card_fun)
-    }
+    card_fun <- reactive({
+      req(setup_code_r(), data_prep_code_r(), computation_model_code_r(), computation(),
+          computation_tbl_imp_code_r(), computation_tbl_eig_code_r(), plot_code_r(), plot_r())
+
+      teal.reporter::report_document(
+
+        "## Setup",
+        teal.reporter::code_chunk(setup_code_r()),
+
+        "## Libraries",
+        teal.reporter::code_chunk(libraries_code_r(), eval = TRUE),
+
+        "## Data Preparations",
+        teal.reporter::code_chunk(data_prep_code_r()),
+
+        "## PCA Model",
+        teal.reporter::code_chunk(computation_model_code_r()),
+
+        "### Principal Components Table",
+        teal.reporter::code_chunk(computation_tbl_imp_code_r()),
+        computation()[["tbl_importance"]],
+
+        "### Eigenvectors Table",
+        teal.reporter::code_chunk(computation_tbl_eig_code_r()),
+        computation()[["tbl_eigenvector"]],
+
+        "### Plot",
+        teal.reporter::code_chunk(
+          plot_code_r() |> styler::style_text() |> paste(collapse = "\n")
+        ),
+        plot_r()
```

Comment:
outputs added to the reportDocument are independent of the code chunk here. It means they are not linked to particular chunk. Implementation in teal.reporter removes all outputs with no exceptions. What if module developer wants to add a plot as an image/table without any code chunk? 

```suggestion
        teal.reporter::code_chunk(
          plot_code_r() |> styler::style_text() |> paste(collapse = "\n"),
          echo = plot_r()
        )
```

Reply:
Yes, at some point I had a linking function but then realized it doesn't give much value.
If you don't want to add any code chunk, then just do not include code in `report_document()`.
If you want to keep some outputs that are R objects, let's add attribute that states they should be kept.

So here are some scenarios below:


1. Code in the report and output but plot only in the output.

```r
teal.reporter::report_document(
  teal.reporter::code_chunk(plot_code),
  plot_object
)
```

2. Code in the report and output and plot in the report and output.

```r
keep_in_report <- function(x, keep = TRUE) {
    attr(x, "keep_in_report") <- keep
    x
}
teal.reporter::report_document(
  teal.reporter::code_chunk(plot_code),
  plot_object |> keep_in_report()
)
```

3. No code in the report and no code in the output and plot in the report and output.

```r
keep_in_report <- function(x, keep = TRUE) {
    attr(x, "keep_in_report") <- keep
    x
}
teal.reporter::report_document(
  plot_object |> keep_in_report()
)
```

4. No code in the report and no code in the output and plot only in the output.

```r
teal.reporter::report_document(
  plot_object
)
```

And cases when you want to have code in the output, but not in the report

```r
keep_in_report <- function(x, keep = TRUE) {
    attr(x, "keep_in_report") <- keep
    x
}
teal.reporter::report_document(
  teal.reporter::code_chunk(plot_code) |> keep_in_report(keep = FALSE),
  plot_object |> keep_in_report()
)
```

Reply:
Also, is you don't want to include code, but just outputs, then this breaks the reproducibility principle. I think we should motivate and influence developers to include as much code as possible, so everything can be reproduced.

And if they don't want to keep the code, `keep_in_report(keep = FALSE)` could do the trick, or you just don't put the code in `report_document()` - it is not evaluated anyway.

Reply:
Nice alternative. Please add it to the teal.reporter. We can discuss this later

Reply:
Alrighty, will try to wrap something for this tomorrow : ) (Friday)

Reply:
Also an example on how to remove things from report but keep in output https://github.com/insightsengineering/teal.reporter/pull/307/commits/707b6bd0631ef134dec150f2a8a4eb59f2763d5e

Reply:
Note: previewer shows all objects, no matter what's their `keep` status.

Reply:
continuation of this thread is here https://github.com/insightsengineering/teal.modules.general/pull/862#discussion_r2057591555